### DISCUSSION TITLE
How to know the type of response body?

### DISCUSSION THREAD

**Discussion by Gumichocopengin8** [tokio-rs/axum#2420] [Q&A] [answered]
### Summary

```rust
Router::new()
        .route("/api/data/:id", get(get_data))

// ...

// I lose type since return type is `impl IntoResponse`
async fn get_data(Path(id): Path<String>) -> impl IntoResponse {
    let data: BodyType = fetch_crate(&id).await;
    Json(data).into_response()
}

#[cfg(test)]
mod tests {
    #[tokio::test]
    async fn should_get_data() {
        let app = app();

        let response = app
            .oneshot(Request::builder().uri("/api/data/test").body(Body::empty()).unwrap())
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);
   
        // I got an error since the type is not correct. How can I get body with the correct type? 
        let body: BodyType = response.into_body().collect().await.unwrap().to_bytes();
        assert_eq!(body.data.size, 3); // I wanna test something like this, so I need to know the type but not sure how to do it
    }
}
```

if I want to test the response body, how can I know the type of JSON response body?
Also, when I use `impl IntoResponse`, the body type info gets lost? 

### axum version

0.7.2

**Comment by davidpdrsn** [comment, marked as answer]
http turns everything into bytes. There is no notion of "preserving types across http". There is just serializing and deserializing bytes.

You have to say more about the specific error you got. Did you mean to serialize using `serde_json`?

**Reply by Gumichocopengin8** [reply]
Yes, I meant to serialize using `serde_json`. I asked this question because [actix-web](https://actix.rs/docs/testing#integration-testing-for-applications) has the function to serialize data with type like this.

```rust
#[cfg(test)]
mod tests {
    use super::*;
    use actix_web::{test, web, App};

    #[actix_web::test]
    async fn test_index_get() {
        let app = test::init_service(
            App::new()
                .app_data(web::Data::new(AppState { count: 4 }))
                .service(index),
        )
        .await;
        let req = test::TestRequest::get().uri("/").to_request();
        let resp: AppState = test::call_and_read_body_json(&app, req).await;

        assert_eq!(resp.count, 4);
    }
}
```
so I was wondering if there's a way to do this in Axum too. Anyways, thanks for answering the question.

**Reply by davidpdrsn** [reply]
axum does not have any built-in test helpers. If you're using `reqwest` you can use its json deserialization helper.

Otherwise you can buffer the body with https://docs.rs/axum/latest/axum/body/fn.to_bytes.html and call https://docs.rs/serde_json/1.0.99/serde_json/fn.from_slice.html
